packages/labs/signals/src/lib/signal-watcher.ts

```diff
@@ -4,23 +4,79 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
-import type {PropertyDeclaration, PropertyValueMap, ReactiveElement} from 'lit';
+import type {ReactiveElement} from 'lit';
 import {Signal} from 'signal-polyfill';
-import {WatchDirective} from './watch.js';
-
-type ReactiveElementConstructor = abstract new (
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  ...args: any[]
-) => ReactiveElement;
 
 export interface SignalWatcher extends ReactiveElement {
-  _updateWatchDirective(d: WatchDirective<unknown>): void;
-  _clearWatchDirective(d: WatchDirective<unknown>): void;
+  _watcher?: Signal.subtle.Watcher;
+}
+
+interface EffectOptions {
+  beforeUpdate?: boolean;
+  manualDispose?: boolean;
+}
+
+export interface ElementEffectOptions extends EffectOptions {
+  element?: SignalWatcher & SignalWatcherApi;
+}
+
+let effectsPending = false;
+const effectWatcher = new Signal.subtle.Watcher(() => {
+  if (effectsPending) {
+    return;
+  }
+  effectsPending = true;
+  queueMicrotask(() => {
+    effectsPending = false;
+    for (const signal of effectWatcher.getPending()) {
+      signal.get();
+    }
+    effectWatcher.watch();
+  });
+});
+
+/**
+ * Executes the provided callback function when any of the signals it accesses
+ * change. If an options object is provided, the `element` property can be used
+ * to specify the element to associate the effect with. The `beforeUpdate`
+ * property can be used to specify that the effect should run before the element
+ * updates. The `manualDispose` property can be used to specify that the effect
+ * should not be automatically disposed when the element is disconnected.
+ *
+ * @param callback
+ * @param options {element, beforeUpdate, manualDispose}
+ * @returns
+ */
+export const effect = (
+  callback: () => void,
+  options?: ElementEffectOptions
+) => {
+  const {element} = options ?? {};
+  if (element === undefined) {
+    const computed = new Signal.Computed(callback);
+    effectWatcher.watch(computed);
+    Signal.subtle.untrack(() => computed.get());
+    return () => {
+      effectWatcher.unwatch(computed);
+    };
+  } else {
+    return element._effect(callback, options);
+  }
+};
+
+interface SignalWatcherApi {
+  _effect(fn: () => void, options?: EffectOptions): () => void;
```

Comment:
Should this be marked `/** @internal */`?

Reply:
Done

Reply:
@justinfagnani if we only run effect for element, should we just expose it directly here? Maybe as `signalEffect` to avoid taking the generic name?

Reply:
Yeah, I think `this.signalEffect()` is ok, thought maybe redundant since "effect" is strongly associated with signals in this world already. `this.updateEffect()` might be ok too? Or does "update" sound like a verb in that usage?

I do think an instance method is fine for this though. As opposed to `effect(this, () => ...)`.

Reply:
changed to `updateEffect`